### PR TITLE
added auto-HPO feature with WandB

### PR DESCRIPTION
**Pull request by yxdyc**
As the title says, related to #47 

- The core modifications are in tools/hpo, and data_juicer/config.py
- The others are from pre-commit auto-modification (isort, yapf, ......)

### CODE REVIEW

File: data_juicer/format/formatter.py [hunk 1]
```diff
@@ -262,4 +262,6 @@ def load_formatter(dataset_path,
 
     # no data
     else:
-        raise NotImplementedError
+        raise ValueError('Can not found local data or huggingface '
+                         'dataset-hub for your given path: '
+                         f'{dataset_path} and suffixes: {suffixes}')
```
Comment (HYLcool) [inline data_juicer/format/formatter.py:267]: This info conflicts with PR #64. Need to be merged later.

File: demos/data_process_hpo/app.py [hunk 2]
```diff
@@ -29,18 +29,24 @@ def setup():
...
+
 def main():
+
     def hello():
 
-        st.image('imgs/data-juicer.png', output_format='png', use_column_width = True)
+        st.image('imgs/data-juicer.png',
+                 output_format='png',
+                 use_column_width=True)
         demo = 'The demo is coming soon😊'
```
Comment (HYLcool) [inline demos/data_process_hpo/app.py:40]: The demo for hpo is in progress. Need to modify this demo after it's prepared.

File: data_juicer/format/formatter.py (hunk 1 above)
Comment (yxdyc) [reply data_juicer/format/formatter.py:267]: Will pay more attention the formatter.py in PR #64

File: demos/data_process_hpo/app.py (hunk 2 above)
Comment (yxdyc) [reply demos/data_process_hpo/app.py:40]: Yes, after accepting this PR, I will prepare demos on MS and HF

File: data_juicer/config/config.py [hunk 3]
```diff
@@ -416,3 +420,66 @@ def display_config(cfg):
...
+    """
+    try:
+        ori_specified_op_names = set()
+        ori_specified_op_idx = {}  # {op_name: op_order}
+        # format of ori_cfg.process
+        # ori_cfg.process[i] = {
+        #     op_in_process_name:
+        #         None if internal_op_para is None else
+        #         namespace_to_dict(internal_op_para)
+        # }
```
Comment (HYLcool) [inline data_juicer/config/config.py:444]: Commented code can be erased.

File: data_juicer/config/config.py [hunk 4]
```diff
@@ -416,3 +420,66 @@ def display_config(cfg):
...
+        for op_order, op_in_process in enumerate(ori_cfg.process):
+            op_name = list(op_in_process.keys())[0]
+            ori_specified_op_names.add(op_name)
+            ori_specified_op_idx[op_name] = op_order
+
+        for new_k, new_v in new_cfg.items():
+            # merge parameters other than `cfg.process` and DJ-OPs
+            if new_k in ori_cfg and new_k != 'process' and '.' not in new_k:
+                print(
```
Comment (HYLcool) [inline data_juicer/config/config.py:453]: It's better to use logger.info instead of pure print.

File: data_juicer/config/config.py [hunk 5]
```diff
@@ -416,3 +420,66 @@ def display_config(cfg):
...
+            else:
+                # merge parameters of DJ-OPs into cfg.process
+                # for nested style, e.g., `remove_table_text_mapper.min_col: 2`
+                key_as_groups = new_k.split('.')
+                if len(key_as_groups) > 1 and \
+                        key_as_groups[0] in ori_specified_op_names:
+                    op_name, para_name = key_as_groups[0], key_as_groups[1]
+                    op_order = ori_specified_op_idx[op_name]
+                    ori_cfg_val = ori_cfg.process[op_order][op_name][para_name]
+                    print(
```
Comment (HYLcool) [inline data_juicer/config/config.py:469]: Same as above.

File: tools/hpo/README.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
+# Hyper-parameter Optimization for Data Recipe
+
+## Auto-HPO
+
+We incorporate an automated HPO tool, WandB [Sweep](https://docs.wandb.ai/guides/sweeps), into Data-Juicer to streamline the finding of good data processing hyper-parameters.
+With this tool, users can investigate correlations and importance scores of
+specific hyper-parameters of data recipes from the HPO view.
+
+*Note*: this is an experimental feature. Auto-HPO for data recipes still has
```
Comment (HYLcool) [inline tools/hpo/README.md:9]: Use ** instread of * to align with other docs.

File: tools/hpo/README.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+a large room to explore. Feel free to provide more suggestions, discussion,
+and contribution via new PRs!
+
+
+## Prerequisite
+You need to install data-juicer first.
+Besides, the tool leverages WandB, install it via `pip install wandb`.
+Before using this tool, you need to run `
+```wandb login``` and enter your WandB
```
Comment (HYLcool) [inline tools/hpo/README.md:18]: There is an extra ` before wandb login

File: tools/hpo/README.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+We provide an illustrative objective "quality_score" in `hpo/objects.py`,
+which uses quality scorer to measure the processed data, and links the average scores to hyper-parameters of data recipes.
+After running it, you will get the result similar to: ![img](https://img.alicdn.com/imgextra/i2/O1CN017fT4Al1bVldeuCmiI_!!6000000003471-2-tps-2506-1710.png)
+
+
+You can implement your own HPO objective in `get_hpo_objective` function, e.g., linking the data
+recipes to
+- model_loss (by replacing the quality scorer into a training procedure),
+- downstream_task (by eplacing the quality scorer into a training and an# Hyper-parameter Optimization for Data Recipe
```
Comment (HYLcool) [inline tools/hpo/README.md:50]: eplacing --> replacing
And the meaning of this sentence is not clear.

File: tools/hpo/README_ZH.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
+# 数据菜谱的自动化超参优化
+
+## Auto-HPO
+
+我们将自动化 HPO (hyper-parameters optimization) 工具 WandB [Sweep](https://docs.wandb.ai/guides/sweeps) 结合到
+Data-Juicer 中，以简化改良数据处理超参数的过程。
+使用此工具，用户可以研究探索 *数据配方的特定超参数* 和 *指定目标度量（如数据质量分、模型loss等）* 之间的 相关性和重要性得分
+
+*注意*：这是一个实验性功能。 用于数据配方的 Auto-HPO 仍然有
```
Comment (HYLcool) [inline tools/hpo/README_ZH.md:9]: Same as English version, * --> **

File: tools/hpo/README_ZH.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+```shell
+# cd tools/hpo
+python execute_hpo.py --config <data-process-cfg-file-path> --hpo_config <hpo-cfg-file-path>
+
+# e.g.,
+python execute_hpo.py --config configs/process.yaml --hpo_config configs/quality_score_hpo.yaml
+```
+
+我们在`hpo/objects.py`中提供了一个示意性的搜索目标 `quality_score`，
+它使用质量评分器来度量处理后的数据，并将平均质分数链接到数据配方的超参数。
```
Comment (HYLcool) [inline tools/hpo/README_ZH.md:42]: Something might be wrong with phrase "平均质分数".

File: tools/hpo/configs/process.yaml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+# Process config example for dataset
```
Comment (HYLcool) [inline tools/hpo/configs/process.yaml:1]: The format of HPO config or how to write a HPO config should be introduced in HPO README files.

File: tools/hpo/configs/quality_score_hpo.yaml [hunk 12]
```diff
@@ -0,0 +1,29 @@
+
```
Comment (HYLcool) [inline tools/hpo/configs/quality_score_hpo.yaml:1]: The format of HPO config or how to write a HPO config should be introduced in HPO README files.

File: tools/hpo/configs/quality_score_hpo.yaml [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  text_length_filter.min_len:
+      distribution: q_log_uniform_values
+      min: 256
+      max: 8192
+
+
+#early_terminate:
+#  type: hyperband
+#  max_iter: 27
+#  s: 2
```
Comment (HYLcool) [inline tools/hpo/configs/quality_score_hpo.yaml:29]: Unused arguments can be removed, or uncomment early termination in this example for better experience.

File: demos/data_process_hpo/app.py (hunk 2 above)
Comment (HYLcool) [reply demos/data_process_hpo/app.py:40]: [TODO item](https://github.com/orgs/alibaba/projects/9/views/4?pane=issue&itemId=43973897) is added in the project.

File: tools/hpo/execute_hpo.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    # 2.2: calculate objective using new hyper-parameters, track the results
+    score = float(object_func(dj_cfg))
+    wandb.log({sweep_configuration['metric']['name']: score})
+
+
+# 3: Start the sweep, iteratively search hyper-parameters
+sweep_id = wandb.sweep(sweep=sweep_configuration,
+                       project=sweep_configuration['sweep_name'])
+
+wandb.agent(sweep_id,
```
Comment (HYLcool) [inline tools/hpo/execute_hpo.py:43]: Logs generated by wandb in the wandb directory should be added to .gitignore

**Comment by yxdyc** [comment]
Fixed the comments by @HYLcool , you can track the modifications in these commits: [this](https://github.com/alibaba/data-juicer/pull/65/commits/2b9763c63c8c4b36346734a1074aaa5c6f1b8c61) and [this](https://github.com/alibaba/data-juicer/pull/65/commits/6293d8d85af9e5dc90dce891d450b1e997a09213).

**Review by HYLcool — APPROVED** [review]
LGTM now.

**Pull request merged**